### ISSUE TITLE
Is it possible to have also non-component files under pages/ directory?

### ISSUE THREAD

**Issue by robertjk** [gatsbyjs/gatsby#26554]
## Summary

Is it really invalid to have a non-component JavaScript file anywhere under `src/pages/`?

## Relevant information

Today I separated some constants from a page component into a separate file, and put that file in the same directory as page component.

### Directory structure

This is the structure of the directory with that page component:

```
src/pages/
|-- page/
    |-- Page.js
    |-- index.js
    |-- someOtherFile.js
```

### File contents

And these are the contents of the files in there:

```
// Page.js

import { A, B, (...) } from "./someOtherFile";

function Page() {
  (...)
}

export default Page;
```
```
// someOtherFile.js

const A = (...);
const B = (...);
(...)

export { A, B, (...) };
```
```
// index.js

import Page from "./Page";
export default Page;
```

### Now build fails

After doing that the build started to fail with the following error message:

```
A page component must export a React component for it to be valid. Please make sure this file exports a React component:

undefined
```

As soon as I remove `someOtherFile.js` the build starts to succeed again. So I assume the problem is that Gatsby treats `someOtherFile.js` as a page component and tries to build a page from it.

### Question

**Is that expected behavior and what I'm doing is simply illegal, or is that a bug that should be reported and fixed?**

**Comment by AdamSiekierski** [comment]
Why would you want to do that? The `pages` directory is meant for only the pages components, if you want to store some utility functions, maybe store them in the `utils` directory?

**Comment by robertjk** [comment]
Why? Because I find that clean architecture - I keep the files close to where they are used. I also do the same for components that are local to some page - are not library components, which are shared between multiple pages. For example:

```
src/pages/
|-- page/
    |-- PageSectionA/
        |-- PageSectionA.js
        |-- index.js
    |-- PageSectionB/
        |-- PageSectionB.js
        |-- index.js
    |-- Page.js
    |-- index.js
```

> maybe store them in the src/utils directory?

I could, but I don't like this approach, as it keeps things far from it's usage and makes the maintenance harder. For example if I want to remove that page completely, in the current architecture, I just need to remove `src/pages/page` directory and I'm done. If I keep `someOtherFile.js` in `src/utils/`, I'll need to remove `src/pages/page/` and `src/utils/someOtherFile.js`. The second is possible to be forgotten/missed in which case we would be left with an unused file in `src/utils/someOtherFile.js`. For the same reason that atchitectural pattern it makes it easier to track unused files.

**Comment by simplesessions** [comment]
Not sure if this helps you, but I tend to handle this by having a `src/features` directory. This allows you to still group functionality that shouldn't be global, but is closely related. This allows a single feature to scale without crowding the global space. And, if something becomes reusable outside of that feature, I promote it to the global `src/components` dir.

e.g.
`src/features/blog/components/Navigation.js`
`src/features/blog/components/Pagination.js`
`src/features/blog/helpers/date.js`

`src/features/sales/components/CheckoutForm.js`
`src/features/sales/components/InterstitialAd.js`
`src/features/sales/styles/theme.js`

**Comment by robertjk** [comment]
> Not sure if this helps you, but I tend to handle this by having a `src/features` directory. This allows you to still group functionality that shouldn't be global, but is closely related. This allows a single feature to scale without crowding the global space. And, if something becomes reusable outside of that feature, I promote it to the global `src/components` dir.
> 
> e.g.
> `src/features/blog/components/Navigation.js`
> `src/features/blog/components/Pagination.js`
> `src/features/blog/helpers/date.js`
> 
> `src/features/sales/components/CheckoutForm.js`
> `src/features/sales/components/InterstitialAd.js`
> `src/features/sales/styles/theme.js`

Yeah, it's a possible solution and thanks for sharing that. I may use exactly the same solution. But first I want to wait for a reply to my question: Is what I'm doing an illegal thing, or a bug that could be addressed and fixed. If it's the first, I will use the same pattern as you. If it's a bug, then I'll just wait for a fix.

**Comment by pieh** [comment]
Gatsby by default adds instance of https://www.gatsbyjs.com/plugins/gatsby-plugin-page-creator/ plugin pointing to `<project_directory>/src/pages` with some things that ARE ignored (so things starting with `_` or `.` or `template-`). You could use one of those conventions to mark some files as in fact not a pages (but still, it's kind of similar to moving files out of `src/pages` - you need to adhere to "hardcoded" convention which doesn't always match the structure of your liking.

But there is a way to add more to the ignore list (thanks to not really documented feature implemented in https://github.com/gatsbyjs/gatsby/pull/17420)

To make use of that feature you would need to add following to your `gatsby-config.js`:

```js
const path = require(`path`)

module.exports = {
  plugins: [
    {
      resolve: `gatsby-plugin-page-creator`,
      options: {
        path: path.join(__dirname, `src/pages`),
        ignore: [`someOtherFile.js`]
      }
    }
  ]
}
```

The `ignore` option is using https://www.npmjs.com/package/micromatch so you could define some custom "globs" if you would use some convention for naming files that are not meant to be pages.

**Comment by robertjk** [comment]
> Gatsby by default adds instance of https://www.gatsbyjs.com/plugins/gatsby-plugin-page-creator/ plugin pointing to `<project_directory>/src/pages` with some things that ARE ignored (so things starting with `_` or `.` or `template-`). You could use one of those conventions to mark some files as in fact not a pages (but still, it's kind of similar to moving files out of `src/pages` - you need to adhere to "hardcoded" convention which doesn't always match the structure of your liking.
> 
> But there is a way to add more to the ignore list (thanks to not really documented feature implemented in #17420)
> 
> To make use of that feature you would need to add following to your `gatsby-config.js` (well, you probably already have some `plugins`, so you would need to merge that with your existing one - the position of page-creator "overwrite" shouldn't matter, so you can append/prepend or put it in the middle):
> 
> ```js
> const path = require(`path`)
> 
> module.exports = {
>   plugins: [
>     {
>       // this overwrites default instance of `page-creator` and adds custom `ignore` options
>       resolve: `gatsby-plugin-page-creator`,
>       options: {
>         // the path is important - it needs to point exactly to `<project-dir>/src/pages`,
>         // if it doesn't, it will then add another instance of the plugin
>         path: path.join(__dirname, `src/pages`),
>         ignore: [`someOtherFile.js`]
>       }
>     }
>   ]
> }
> ```
> 
> The `ignore` option is using https://www.npmjs.com/package/micromatch so you could define some custom "globs" if you would use some convention for naming files that are not meant to be pages.

Thanks. That's a useful information.

That assures me that Gatsby is meant to have only page component files in `src/pages/` directory. I'll do what @simplesessions suggested, which is moving `src/pages/` to `src/features/` and importing those components in files in `src/pages/`:

```
src/
|-- pages/
|   |-- page-1.js
|   |-- page-2.js
|   (...)
|
|-- features/
|   |-- Page1/
|   |    |-- Page1.js
|   |    |-- index.js
|   |
|   |-- Page2/
|   |    |-- Page1.js
|   |    |-- index.js
|   |
|   (...)
```

```javascript
// src/pages/page-1.js
export { default } from "~features/Page1";
```

```javascript
// src/features/Page1/Page1.js

function Page1() {
  (...)
}

export default Page1;
```

```javascript
// src/features/Page1/index.js
export { default } from "./Page1";
```

**Comment by robertjk** [comment]
Having separate directory `src/features/` (or any other name you'd prefer) seems like the best solution for my needs. The reason is that for Gatsby is an unknown directory and so Gatsby doesn't put any restrictions on it's structure. Even if you add `gatsby-ignore` directive, or something similar, `src/pages/` directory will still have some other restrictions or side-effects, e.g. any nested directory will create a new URL route. The only way to "be free" and have things organized by page/feature seems to maintain another directory. And things can still stay very clean and well organized, if I keep the `src/pages/` directory flat and doing nothing more than import and export components from `src/features/`.

**Comment by kingluddite** [comment]
I'm trying to follow this pattern but how do I deal with page level GraphQL? If I keep in at the page level `src/pages/about.js` I get this error "any GraphQL fragments or queries in this file were not processed." 

I move it into the `features/AboutPage.js` and get the same error. I tried to change it to a Static Query and the page did not render. Any ideas or guidance?

```
export const query = graphql`
  query AboutPageQuery {
    aboutPage: sanityPage(name: { eq: "About" }) {
      name
      headingOne
      expertise
        bioTextBlock: _rawContent
      }
    }
  }
`
```

**Comment by barbalex** [comment]
I have been adding additional files just as @robertjk wanted to.

Weird things have happened since. One of which: updating things in these files sometimes did not change what dev was presenting. I tried multiple things to get the dev ui to update:
- restart dev
- restart the pc
- `gatsby clean`

Nothing helped.

Until I found this issue and renamed my files to begin with underscores.

Maybe this can help others encountering the weird problems.
